platform-sdk/docs/proposals/healthMonitor/healthMonitorProposal.md

```diff
@@ -0,0 +1,127 @@
+is 100s. If the second queue clears its backlog and becomes healthy, it will then report that the system unhealthiness
+factor is 10s.
+
+The health monitor publishes health reports over a wire. If it sends a duration of 0s, that means that the system is in
+a healthy state. As the system becomes more and more unhealthy, it will periodically publish higher and higher
+durations.
+
+#### Reaction
+
+Various subsystems react to the health monitor's reports in different ways.
+
+##### Event Creation
+
+If the health monitor reports an unhealthy duration exceeding X seconds (configurable), then the creation of self
+events is disabled. Once the system reports an unhealthy duration below X seconds, event creation may resume.
+
+##### Gossip
+
+The sync gossip algorithm has the concept of a "sync permit". In order to initiate a sync operation, a permit must
+be acquired. When the sync is completed, the permit is released. If there are 5 permits available (configurable),
+then at any point in time there may be at most 5 sync operations in progress.
+
+When the health monitor reports that the system is becoming healthy, the gossip system takes no action until the
+unhealthy duration exceeds the grace period (a configurable duration). This is designed to prevent the gossip
+subsystem from reacting too aggressively to little blips.
+
+When the system is unhealthy for a duration that exceeds the grace period, the gossip system will begin revoking
+permits. Revoking a permit does not halt a sync operation that is currently in progress, but it can prevent a new
+sync operation from starting. For each second that the system is unhealthy beyond the grace period, permits are revoked
+at the rate specified by configuration.
+
+When the system eventually becomes healthy again, permits are gradually returned according to the rate specified in
+configuration.
+
+There is a special edge case when permits are being returned. Suppose the system has been unhealthy for a while, and
+then suddenly the backlog of work is cleared and the system becomes healthy. If all permits have been revoked, then
+it may take several seconds for the first permit to be returned. This means that a healthy system may have zero permits
+for a time after becoming healthy. To combat this problem, there is a configurable "minimum healthy permits" value.
+When the system is healthy, this is the minimum number of un-revoked permits that will be available. Note that this
+value is not respected in an unhealthy system, which should eventually have 0 permits if the problem is not resolved.
+
+##### Acceptance of application transactions
+
+If the health monitor reports an unhealthy duration exceeding X seconds (configurable), then the acceptance of
+application transactions is disabled. Once the system reports an unhealthy duration below X seconds,
+application transactions are accepted again.
```

Comment:
Maybe this could be changed to just say, if the node is unhealthy, then it won't accept new transactions. We should elsewhere define unhealthy as a boolean toggled based on some criteria. It could be just duration exceeding X seconds, or could be based on exponential backoff, etc, but this part of the specification doesn't need to define how we compute whether the node is healthy or not, we just need to say, when it is unhealthy, we don't accept new transactions. We should also describe why this is the case. I think the answer is, because we don't want to create a self-event, or because we won't be creating self-events and this will cause transactions to timeout.

Reply:
This entire paradigm is made significantly simpler if we can make the following paradigm shift for event/creation & transaction acceptance. My goal for the health monitor change was to shift the current paradigm as little as possible, with the idea being that we should address this issue as a separate change.

# Current Algorithm

There is a method the app can call to send transactions. This puts the transactions into a queue. When the platform creates an event, it takes transactions from the queue. If the queue is full or if the platform thinks now is a bad time, it rejects transactions (the method returns a boolean to indicate if it was accepted).

# Proposed Algorithm

App registers a provider with the platform. When the platform is about to create an event, it requests up to X bytes of transactions from the app. If the app has transactions available, it provides those transactions.

The platform exposes various health factors through an API to the app. It is up to the app to determine the policy for when a user transaction should be accepted or not.